### ISSUE TITLE
[LSR] Wrong code at -O1 on x86_64-linux_gnu since Clang-13

### ISSUE THREAD

**Issue by shao-hua-li** [llvm/llvm-project#62515]
```console
$ clang-tk -v
clang version 17.0.0 (https://github.com/llvm/llvm-project.git 4c2d29f2fc78782a469ccff625dd29d66d4d9f9b)
Target: x86_64-unknown-linux-gnu
Thread model: posix
InstalledDir: /zdata/shaoli/compilers/ccbuilder-compilers
Found candidate GCC installation: /usr/lib/gcc/x86_64-linux-gnu/10
Found candidate GCC installation: /usr/lib/gcc/x86_64-linux-gnu/11
Found candidate GCC installation: /usr/lib/gcc/x86_64-linux-gnu/9
Selected GCC installation: /usr/lib/gcc/x86_64-linux-gnu/11
Candidate multilib: .;@m64
Selected multilib: .;@m64
$
$ clang-tk -O0 a.c && ./a.out
0
$ clang-tk -fsanitize=address,undefined a.c && ./a.out
0
$ clang-tk -O1 a.c && ./a.out
Floating point exception
$ clang-12 -O1 a.c && ./a.out
0
$
$ cat a.c
int printf(const char *, ...);
long a;
static int b, d, e;
char g, h;
int i(int j) {
  int c = 0;
  while (c < 4 && 5 << c < j)
    c++;
  return c;
}
int k(int j) {
  int f = 0;
  if (j < 0)
    f++;
  do {
    f++;
    j /= 10;
  } while (j);
  return f;
}
long(l)(long j) { return j == 0 ?0: a / j; }
void m(int j) {
  d = k(j - 54);
  h = l(d - 3);
}
int main() {
  int *n[] = {&b, &b};
  e = i(b - 40);
  m(-e);
  printf("%d\n", g);
}
$
```

Compiler explorer: https://godbolt.org/z/9ecEKnGbh

`opt-bisect-limit` suggests the issue is in `Loop Strength Reduction on loop`

**Comment by fhahn** [comment]
Looks like this may be X86 specific. I can reproduce it on X86 macOS but not AArch64 macOS. The IR after the middle end is the same in both cases.

**Comment by shao-hua-li** [comment]
A simpler test case:

Compiler explorer: https://godbolt.org/z/cPaosha6h

```C
int printf(const char *, ...);
int a, b;
unsigned c(int d, int e) { return (d & 3) << 3 | e & 7; }
int main() {
  b = 0;
  for (; c(b + 35, b + 31) - 31 <= -11; b = b - 2) {
    int *f = &a;
    *f ^= 2;
  }
  printf("%d\n", a);
}
```

**Comment by fhahn** [comment]
The original reproduce should be fixed by https://reviews.llvm.org/D153004

**Comment by fhahn** [comment]
The simpler case is fixed by the patch for #62660 so I am going to close this one as the original issue has been fixed.
